README.md
# OXID Console Moduleactivator

Custom command for the [OXID console](https://docs.oxid-esales.com/developer/en/6.2/development/tell_me_about/console.html).

## Installation

`composer require proudcommerce/oxid-console-moduleactivator`

## Commands

pc:module:activator
--------------------

* Description: Activates multiple modules, based on a YAML file.

* Usage:

`vendor/bin/oe-console pc:module:activator [-s|--skipDeactivation] [-d|--clearModuleData] [--] <yaml-file>`

* Example (activate all modules defined in the YAML file basedon a white- or blacklist):

`vendor/bin/oe-console pc:module:activator -s ../modules.yml`

### YML file

```yaml
whitelist:
  1:
    - ocb_cleartmp
    - moduleinternals
    - custommodule1
    #- ddoevisualcms
    #- ddoewysiwyg
  2:
    - ocb_cleartmp
    - custommodule2
priorities:
  1:
    moduleinternals:
        1200
    ocb_cleartmp:
        950
installations:
    - source/modules/myvendor/custommodule1
    - source/modules/myvendor/custommodule2
```

Supports either a __"whitelist"__ or a __"blacklist"__ entry with multiple shop ids and the desired module ids to activate (whitelist) or to exclude from activation (blacklist).

If you want, you can also specify __a YAML string on the command line instead of a file__, e.g.:

```bash
oe-console pc:module:activator $'whitelist:
  1:
    - oepaypal
' --shop-id=1
```

### Module priorities

With __"priorities"__, you can define the order (per subshop) in which the modules will be activated. You can defined priorities per subshop, see example above.

### Installing modules prior to activation

By specifying an __"installations"__ array, you can make sure that the modules are really "installed" in the shop (modules can be installed via Composer or by calling e.g. `vendor/bin/oe-console oe:module:install-configuration source/modules/oxcom/ocbcleartmp`). If a module of that list is not yet installed, `oe:module:install-configuration` will be called (please note that you still have to __add the Composer namespace for autoloading__ in the main _composer.json_ e.g.). So this is an "edge-case", if you add e.g. custom project specific modules directly into the _"source/modules/"_ folder and add them manually to the main _composer.json_ (so you don't need to have a separate Composer repository for each module):

```json
    "autoload": {
        "psr-4": {
            "MyVendor\\CustomModule1\\": "./source/modules/myvendor/custommodule1",
            "MyVendor\\CustomModule2\\": "./source/modules/myvendor/custommodule2"
        }
    }
```

A __better alternative__ might be to use the ["wikimedia/composer-merge-plugin"](https://github.com/wikimedia/composer-merge-plugin) plugin, so that your module's _composer.json_ will be used directly.

### Arguments

**yaml:**

* Name: yaml
* Description: YAML module list filename or YAML string. The file path is relative to the shop "source/" path

### Options:

**skipDeactivation:**

* Name: `--skipDeactivation`
* Shortcut: `-s`
* Accept value: no
* Is value required: no
* Description: Skip deactivation of modules, only activate.
* Default: `false`

**clearModuleData:**

* Name: `--clearModuleData`
* Shortcut: `-d`
* Accept value: no
* Is value required: no
* Description: Clear module data in oxconfig table.
* Default: `false`


CHANGELOG.md
# Change log for ProudCommerce OXID console moduleactivator

All notable changes to this project will be documented in this file.
The format is based on [Keep a Changelog](http://keepachangelog.com/)
and this project adheres to [Semantic Versioning](http://semver.org/).

## [v1.1.3] - 2022-09-30

### Added

- made compatible with OXID 6.5 series and Symfony Component from version 5.0

## [v1.1.2] - 2020-05-26

### Changed

- improve console output

## [v1.1.1] - 2020-05-25

### Fixed

- clear modules in db (shopid)

## [v1.1.0] - 2020-05-22

### Added

- "installations" array in YAML config to "install" modules in OXID

## [v1.0.2] - 2020-05-21

### Fixed

- remove deprecated OXID code
- changed some function visibility

## [v1.0.1] - 2020-05-16

### Fixed

- add missing services.yaml stuff ;-)


## [v1.0.0] - 2020-05-16

- Release ;-)


src/Command/ModuleactivatorCommand.php
<?php declare(strict_types=1);

namespace ProudCommerce\OxidConsoleModuleactivator\Command;

use OxidEsales\EshopCommunity\Core\Registry;
use OxidEsales\EshopCommunity\Internal\Framework\Module\Configuration\Dao\ShopConfigurationDaoInterface;
use OxidEsales\EshopCommunity\Internal\Framework\Module\Setup\Service\ModuleActivationServiceInterface;
use OxidEsales\EshopCommunity\Internal\Framework\Module\State\ModuleStateServiceInterface;
use OxidEsales\EshopCommunity\Internal\Transition\Utility\ContextInterface;
use OxidEsales\EshopCommunity\Internal\Framework\Database\QueryBuilderFactoryInterface;
use OxidEsales\EshopCommunity\Internal\Framework\Module\Install\Service\ModuleConfigurationInstallerInterface;
use Symfony\Component\Console\Command\Command;
use Symfony\Component\Console\Input\InputArgument;
use Symfony\Component\Console\Input\InputInterface;
use Symfony\Component\Console\Input\InputOption;
use Symfony\Component\Console\Output\OutputInterface;
use Symfony\Component\Yaml\Yaml;
use Symfony\Component\Console\Input\ArrayInput;
use Webmozart\PathUtil\Path;

class ModuleactivatorCommand extends Command
{

    /**
     * Module activation priorities
     *
     * @var array
     */
    private $aPriorities = [];
    /**
     * Shop ids in YAML
     *
     * @var array
     */
    private $aYamlShopIds = [];

    /**
     * @var ShopConfigurationDaoInterface
     */
    private $shopConfigurationDao;

    /**
     * @var ContextInterface
     */
    private $context;

    /**
     * @var ModuleActivationServiceInterface
     */
    private $moduleActivationService;
    /**
     * @var ModuleStateServiceInterface
     */
    private $stateService;
    /**
     * @var QueryBuilderFactoryInterface
     */
    private $queryBuilderFactory;
    /**
     * @var ModuleConfigurationInstallerInterface
     */
    private $moduleConfigurationInstaller;

    /**
     * @param ShopConfigurationDaoInterface         $shopConfigurationDao
     * @param ContextInterface                      $context
     * @param ModuleActivationServiceInterface      $moduleActivationService
     * @param ModuleStateServiceInterface           $stateService
     * @param QueryBuilderFactoryInterface          $queryBuilderFactory
     * @param ModuleConfigurationInstallerInterface $moduleConfigurationInstaller
     */
    public function __construct(
        ShopConfigurationDaoInterface $shopConfigurationDao,
        ContextInterface $context,
        ModuleActivationServiceInterface $moduleActivationService,
        ModuleStateServiceInterface $stateService,
        QueryBuilderFactoryInterface $queryBuilderFactory,
        ModuleConfigurationInstallerInterface $moduleConfigurationInstaller
    )
    {
        parent::__construct(null);

        $this->shopConfigurationDao = $shopConfigurationDao;
        $this->context = $context;
        $this->moduleActivationService = $moduleActivationService;
        $this->stateService = $stateService;
        $this->queryBuilderFactory = $queryBuilderFactory;
        $this->moduleConfigurationInstaller = $moduleConfigurationInstaller;
    }

    /**
     * Command configuration
     *
     * @return void
     */
    protected function configure()
    {
        $this->setName('pc:module:activator')
            ->setDescription('Activates multiple modules, based on a YAML file')
            ->addOption('skipDeactivation', 's', InputOption::VALUE_NONE, "Skip deactivation of modules, only activate.")
            ->addOption('clearModuleData', 'd', InputOption::VALUE_NONE, "Clear module data in oxconfig.")
            ->addArgument('yaml', InputArgument::REQUIRED, 'YAML module list filename or YAML string. The file path is relative to the shop installation_root_path');

        $help = <<<HELP
<info>usage:</info>
<comment>oe-console pc:module:activator modules.yml</comment>
- to activate all modules defined in the YAML file based
on a white- or blacklist

Example:

```yaml
whitelist:
1:
- ocb_cleartmp
- moduleinternals
#- ddoevisualcms
#- ddoewysiwyg
2:
- ocb_cleartmp
priorities:
1:
moduleinternals:
    1200
ocb_cleartmp:
    950
```

Supports either a __"whitelist"__ or a __"blacklist"__ entry with multiple shop ids and the desired module ids to activate (whitelist) or to exclude from activation (blacklist).

With "priorities", you can define the order (per subshop) in which the modules will be activated.

If you want, you can also specify __a YAML string on the command line instead of a file__, e.g.:

```bash
oe-console pc:module:activator $'whitelist:\n  1:\n    - oepaypal\n' --shop-id=1
```
HELP;
        $this->setHelp($help);
    }

    /**
     * Executes the current command.
     *
     * @param InputInterface  $input  An InputInterface instance
     * @param OutputInterface $output An OutputInterface instance
     */
    protected function execute(InputInterface $input, OutputInterface $output)
    {
        $activateShopId = $this->context->getCurrentShopId();
        $output->writeLn("<info>START module activator shop " . $activateShopId . "</info>");
        $clearModuleData = $input->getOption('clearModuleData');
        if ($clearModuleData) {
            $output->writeLn("<info>Clearing module data in DB!</info>");
            $this->clearModuleData($activateShopId);
        }
        $skipDeactivation = $input->getOption('skipDeactivation');
        $shopConfiguration = $this->shopConfigurationDao->get(
            $this->context->getCurrentShopId()
        );

        // now try to read YAML
        $moduleYml = $this->getYaml($input->getArgument('yaml'));
        $moduleValues = Yaml::parse($moduleYml);
        if ($moduleValues && is_array($moduleValues)) {
            // check if we also have to install modules first
            $this->installModules($moduleValues, $input, $output);
            $this->aPriorities = $this->getPriorities($moduleValues, $input, $output);
            // use whitelist
            if (isset($moduleValues['whitelist'])) {
                $this->aYamlShopIds = array_keys($moduleValues['whitelist']);
                foreach ($moduleValues['whitelist'] as $shopId => $moduleIds) {
                    if ($activateShopId && $activateShopId != $shopId) {
                        $output->writeLn("<comment>Skipping shop '$shopId'!</comment>");
                        continue;
                    }

                    if (count($this->aPriorities)) {
                        $output->writeLn("<comment>Orig module order:</comment>" . print_r($moduleIds, true));
                        uasort($moduleIds, [$this, "sortModules"]);
                        $output->writeLn("<comment>Sorted module order:</comment>" . print_r($moduleIds, true));
                    }

                    foreach ($moduleIds as $moduleId) {
                        if (!$this->isInstalled($moduleId)) {
                            $output->writeLn('<error>Module not found: ' . $moduleId . '</error>');
                            continue;
                        }
                        // activate
                        if (!$skipDeactivation) {
                            if ($this->stateService->isActive($moduleId, $shopId) === true) {
                                $this->moduleActivationService->deactivate($moduleId, $shopId);
                                $output->writeLn("<info>Module '$moduleId' deactivated</info>");
                            } else {
                                $output->writeLn("<comment>Module '$moduleId' not active</comment>");
                            }
                        }
                        if ($this->stateService->isActive($moduleId, $shopId) === false) {
                            $this->moduleActivationService->activate($moduleId, $shopId);
                            $output->writeLn("<info>Module '$moduleId' activated</info>");
                        } else {
                            $output->writeLn("<comment>Module '$moduleId' already active</comment>");
                        }
                    }
                }
            } elseif (isset($moduleValues['blacklist'])) {
                // use blacklist
                $this->aYamlShopIds = array_keys($moduleValues['blacklist']);
                $aModules = $shopConfiguration->getModuleConfigurations();

                if (count($this->aPriorities)) {
                    $output->writeLn("<comment>Orig module order:</comment>" . print_r(array_keys($aModules), true));
                    uasort($aModules, [$this, "sortModules"]);
                    $output->writeLn("<comment>Sorted module order:</comment>" . print_r(array_keys($aModules), true));
                }

                foreach ($aModules as $moduleConfiguration) {
                    $moduleId = $moduleConfiguration->getId();
                    foreach ($moduleValues['blacklist'] as $shopId => $moduleIds) {
                        if ($activateShopId && $activateShopId != $shopId) {
                            $output->writeLn("<comment>Skipping shop '$shopId'!</comment>");
                            continue;
                        }

                        if (in_array($moduleId, $moduleIds)) {
                            $output->writeLn("<comment>Module blacklisted: '$moduleId' - skipping!</comment>");
                            continue 2;
                        }
                        if (!$this->isInstalled($moduleId)) {
                            $output->writeLn('<error>Module not found: ' . $moduleId . '</error>');
                            continue 2;
                        }
                        // activate
                        if (!$skipDeactivation) {
                            if ($this->stateService->isActive($moduleId, $shopId) === true) {
                                $this->moduleActivationService->deactivate($moduleId, $shopId);
                                $output->writeLn("<info>Module '$moduleId' deactivated</info>");
                            } else {
                                $output->writeLn("<comment>Module '$moduleId' not active</comment>");
                            }
                        }
                        if ($this->stateService->isActive($moduleId, $shopId) === false) {
                            $this->moduleActivationService->activate($moduleId, $shopId);
                            $output->writeLn("<info>Module '$moduleId' activated</info>");
                        } else {
                            $output->writeLn("<comment>Module '$moduleId' already active</comment>");
                        }
                    }
                }
            } else {
                $output->writeLn("<comment>No modules to activate for subshop " . $activateShopId . "!</comment>");
            }
        } else {
            $output->writeLn("<error>No valid YAML data found!</error>");
        }
        $output->writeLn("<info>END module activator shop " . $activateShopId . "</info>");

        return Command::SUCCESS;
    }

    /**
     * Get YAML string, either from file or from string
     *
     * @param string $ymlString The relative file path, from shop INSTALLATION_ROOT_PATH/oxrun_config/ OR a YAML string
     * @param string $basePath  Alternative root dir path, if a file is used
     *
     * @return string
     */
    private function getYaml($ymlString, $basePath = '')
    {
        // is it a file?
        if (strpos(strtolower($ymlString), '.yml') !== false
            || strpos(strtolower($ymlString), '.yaml') !== false
        ) {
            if ($basePath == '') {
                $basePath = $this->getConfigPath();
            }
            $ymlFile = $basePath . $ymlString;
            if (file_exists($ymlFile)) {
                $ymlString = file_get_contents($ymlFile);
            }
        }

        return $ymlString;
    }

    /**
     * Get OXID source path
     *
     * @return string
     */
    private function getConfigPath()
    {
        return Registry::getConfig()->getConfigParam('sShopDir') . DIRECTORY_SEPARATOR;
    }

    /**
     * @param string $moduleId
     *
     * @return bool
     */
    private function isInstalled(string $moduleId): bool
    {
        $shopConfiguration = $this->shopConfigurationDao->get(
            $this->context->getCurrentShopId()
        );

        return $shopConfiguration->hasModuleConfiguration($moduleId);
    }

    /**
     * Sort modules by priority descending per subshop
     *
     * @param Module $a
     * @param Module $b
     *
     * @return int
     */
    public function sortModules($a, $b)
    {
        $aP = $bP = 0;
        // we may have module ids in whitelist
        if (is_string($a) && is_string($b)) {
            $aID = $a;
            $bID = $b;
        } else {
            // or Module objects if using blacklist
            $aID = $a->getId();
            $bID = $b->getId();
        }
        foreach ($this->aYamlShopIds as $shopId) {
            // check if subshop priorities defined
            if (isset($this->aPriorities[ $shopId ])) {
                if (isset($this->aPriorities[ $shopId ][ $aID ])) {
                    $aP = $this->aPriorities[ $shopId ][ $aID ];
                }
                if (isset($this->aPriorities[ $shopId ][ $bID ])) {
                    $bP = $this->aPriorities[ $shopId ][ $bID ];
                }
            }
        }
        //die($aID . ' - ' . $bID . ' - ' . $aP . ' - ' . $bP);
        if ($aP == $bP) {
            return 0;
        }

        return ($aP > $bP) ? -1 : 1;
    }

    /**
     * Get module priorities, if any
     *
     * @param array           $moduleValues Yaml entries as array
     * @param InputInterface  $input        An InputInterface instance
     * @param OutputInterface $output       An OutputInterface instance
     *
     * @return array
     */
    private function getPriorities($moduleValues, $input, $output)
    {
        $aPriorities = [];
        $activateShopId = $this->context->getCurrentShopId();
        if (isset($moduleValues['priorities'])) {
            foreach ($moduleValues['priorities'] as $shopId => $modulePrios) {
                if ($activateShopId && $activateShopId != $shopId) {
                    continue;
                }
                $aPriorities[ $shopId ] = $modulePrios;
            }
        }
        if (count($aPriorities)) {
            $output->writeLn("<comment>Module Priorities:</comment>");
            $output->writeLn(print_r($aPriorities, true));
        }

        return $aPriorities;
    }

    /**
     * Get module installations, if any, and install them
     *
     * @param array           $moduleValues Yaml entries as array
     * @param InputInterface  $input        An InputInterface instance
     * @param OutputInterface $output       An OutputInterface instance
     *
     * @return array
     */
    private function installModules($moduleValues, $input, $output)
    {
        /* @var Symfony\Component\Console\Application $app */
        $app = $this->getApplication();
        if (isset($moduleValues['installations'])) {
            foreach ($moduleValues['installations'] as $moduleDir) {
                $sourceDir = $moduleDir;
                // target dir is optional, may be specified after ':' separator,
                $targetDir = '';
                if (strpos($moduleDir, ':') !== false) {
                    $sourceTarget = explode(':', $moduleDir);
                    $sourceDir = $sourceTarget[0];
                    $targetDir = $sourceTarget[1];
                }
                $output->writeLn("<comment>Checking if module {$sourceDir} is installed ...</comment>");
                $fullPath = $this->getAbsolutePath($sourceDir);
                $fullTargetPath = $this->getAbsolutePath($targetDir);
                if (!$this->moduleConfigurationInstaller->isInstalled($fullPath)) {
                    // module configuration installer doesn't create folders
                    if (!file_exists($fullTargetPath)) {
                        try {
                            mkdir($fullTargetPath, 0775, true);
                        } catch (\Exception $ex) {
                            $output->writeLn("<error>Unable to create folder {$fullTargetPath}!</error>");
                        }
                    }
                    $arguments = [
                        'command'            => 'oe:module:install-configuration',
                        'module-source-path' => $sourceDir,
                        'module-target-path' => $targetDir
                    ];
                    $activateInput = new ArrayInput($arguments);
                    $output->writeLn("<comment>Installing module {$sourceDir} ...</comment>");
                    $app->find('oe:module:install-configuration')->run($activateInput, $output);
                } else {
                    $output->writeLn("<comment>Module {$sourceDir} is already installed!</comment>");
                }
            }
        }
    }

    /**
     * @param string $path
     *
     * @return string
     */
    private function getAbsolutePath(string $path): string
    {
        return Path::isRelative($path)
            ? Path::makeAbsolute($path, getcwd())
            : $path;
    }

    /**
     * Delete module entries from oxconfig table
     *
     * @param int $shopId
     *
     * @return void
     */
    private function clearModuleData($shopId = false)
    {
        $aVarnames = [
            'aDisabledModules',
            'aLegacyModules',
            'aModuleFiles',
            'aModulePaths',
            'aModules',
            'aModuleTemplates'
        ];
        $queryBuilder = $this->queryBuilderFactory->create();
        $queryBuilder
            ->delete('oxconfig')
            ->where("oxvarname IN('" . implode("','", array_values($aVarnames)) . "')")
            ->andWhere("oxshopid = " . $shopId)
            ->execute();
    }
}
